compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64StringUTF16CompressOp.java

```diff
@@ -0,0 +1,185 @@
+            Register count = scratchReg1.getRegister();
+            Register currData = scratchReg2.getRegister();
+
+            masm.mov(64, srcAddress, asRegister(src));
+            masm.mov(64, destAddress, asRegister(dst));
+            masm.lsl(64, count, asRegister(len), 1);
+            masm.mov(largestLatin1, 0xFF);
+
+            masm.align(16);
+            masm.bind(loop);
+            /*
+             * If the current char is greater than 0xFF, stop compressing and return zero as result
+             */
+            masm.ldr(16, currData, AArch64Address.createImmediateAddress(16, AddressingMode.IMMEDIATE_POST_INDEXED, srcAddress, 2));
+            masm.cmp(64, currData, largestLatin1);
+            masm.branchConditionally(ConditionFlag.GT, failToCompress);
+
+            masm.str(8, currData, AArch64Address.createImmediateAddress(8, AddressingMode.IMMEDIATE_POST_INDEXED, destAddress, 1));
+            masm.subs(64, count, count, 2);
+            masm.branchConditionally(ConditionFlag.GT, loop);
+        }
+    }
+
+    private void emitSIMD(AArch64MacroAssembler masm, Label failToCompress, Label ret) {
+        Register srcChunkAddress = asRegister(temp1);
+        Register destChunkAddress = asRegister(temp2);
+        Register chunkPart1RegV = asRegister(vectorTemp1);
+        Register chunkPart2RegV = asRegister(vectorTemp2);
+        Register tmpRegV1 = asRegister(vectorTemp3);
+
+        Label simdLoop = new Label();
+
+        AArch64ASIMDAssembler.ElementSize eSize = AArch64ASIMDAssembler.ElementSize.fromSize(16);
+        try (ScratchRegister scratchRegister1 = masm.getScratchRegister(); ScratchRegister scratchRegister2 = masm.getScratchRegister()) {
+            Register lastChunkAddress = scratchRegister1.getRegister();
+            Register endOfSrcAddress = scratchRegister2.getRegister();
+            masm.mov(64, srcChunkAddress, asRegister(src));
+            masm.mov(64, destChunkAddress, asRegister(dst));
+            masm.add(64, endOfSrcAddress, srcChunkAddress, asRegister(len), AArch64Assembler.ShiftType.LSL, 1);
+            masm.sub(64, lastChunkAddress, endOfSrcAddress, chunkByteSize);
+
+            masm.align(16);
+            masm.bind(simdLoop);
+            masm.fldp(128, chunkPart1RegV, chunkPart2RegV, AArch64Address.createImmediateAddress(128, AddressingMode.IMMEDIATE_PAIR_POST_INDEXED, srcChunkAddress, 32));
+            /*
+             * If a char in the chunk is greater than 0xFF, stop compressing and return zero as
+             * result
+             */
+            masm.neon.orrVVV(AArch64ASIMDAssembler.ASIMDSize.FullReg, tmpRegV1, chunkPart1RegV, chunkPart2RegV);
+            masm.neon.uzp2VVV(AArch64ASIMDAssembler.ASIMDSize.FullReg, eSize.narrow(), tmpRegV1, tmpRegV1, tmpRegV1);
+            masm.fcmpZero(64, tmpRegV1);
+            masm.branchConditionally(ConditionFlag.NE, failToCompress);
```

Comment:
Don't you need to set all of the values in dst up until the first value which is larger than 1 byte?

Reply:
Done